### PR TITLE
Cip10 commit

### PR DESCRIPTION
**Pull request by chiguireitor**

### CODE REVIEW

File: cip-0010.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+--------|-------------|---|-------------|---------------------|---------------|---|---------------------|---------------|---------------------|
+ 16 Bits|   21 bytes  |...|   21 bytes  |         1-bit       |   AssetList   |...|        1-bit        |   AssetList   |        1-bit        |
+
+`AssetList` is built in the following way:
+
+Asset Name|# Z of recipients|LUT Idx 0|Amount Idx 0|...|LUT Idx Z|Amount Idx Z|
+----------|-----------------|---------|------------|---|---------|------------|
+   INT    |     X bits      |  X bits |     INT    |...|  X bits |     INT    |
+
+Note that addresses are only 21 bytes. That's the result of decoding the base58 addresses into binary representation and stripping the last 4 bytes which are just a checksum.
```
Comment (dexX7) [inline cip-0010.md]: > That's the result of decoding the base58 addresses into binary representation and stripping the last 4 bytes which are just a checksum.

Are you suggesting to convert the base58 address into it's actual binary representation, or use the underlying RIPEMD-160 hash with network prefix?

File: cip-0010.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+# Definitions
+
+ * **LUT**: A lookup table. Elements are expressed as indexes into this table with the minimum amount of bits required for the total amount of elements in it.
+ * **RLE**: Run-length Encoding, a compression technique used to express multiple values grouped by type. A run-length specifies a group value and amount of elements contained in that group.
+ * **INT**: A big endian 8 byte integer,
+
+# Overview
+
+The main idea is to create a LUT for addreses so they can be expressed by simple indexes, and then compress each amount sent with simple RLE. This would save space by reducing the amount of bits dedicated to expressing the
+addresses. In the most extreme cases, reduction of bits needed to express each address is 168 times than
```
Comment (deweller) [inline cip-0010.md]: ...168 times *smaller* than...

File: cip-0010.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+#### Example
+
+The following example is broken into several newlines, but the data stream is continuous on the encoded TX data. Numbers are preceded by a marker, 'x' for hexadecimal numbers, 'b' for bits.
+
+ * `x2000` LUT Size: 2 addresses, 1 bit to represent each one
+ * `x00F3A6B6E4A093E5A5B9DA76977A5270FD4D62553E` Base58 decoded address (into hex, 21 bytes), LUT index 0
+ * `x009BDE9D244E7ABD061B0DCC1DFA8C0D880CD7DA21` Base58 decoded address (into hex, 21 bytes), LUT index 1
+ * `b1` Bit, Valid asset after this
+ * `x0100000000000000` 64 bits XCP asset ID
```
Comment (deweller) [inline cip-0010.md]: Likewise, isn't this `0x0000000000000001`?

File: cip-0010.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ * `b1` 2 recipients
+ * `b0` Address 0 into the lookup table
+ * `x0000000001000000` Send 1 of the current Asset (XCP, divisible)
```
Comment (deweller) [inline cip-0010.md]: 1 should be `0x0000000005f5e100`

File: cip-0010.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+ * `b1` Address 1 into the lookup table
+ * `x0000000004000000` Send 4 of the current Asset (XCP)
+ * `b0` End sending data
+
+The whole transaction would then be encoded as (padded with 0s on the last byte):
+
+`020000f3a6b6e4a093e5a5b9da76977a5270fd4d62553e009bde9d244e7abd061b0dcc1dfa8c0d880cd7da21800000000000000000c0000020000000001000000040000000`
```
Comment (deweller) [inline cip-0010.md]: Please update after the previous corrections are made.

File: cip-0010.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+`AssetList` is built in the following way:
+
+Asset Name|# Z of recipients|LUT Idx 0|Amount Idx 0|...|LUT Idx Z|Amount Idx Z|
+----------|-----------------|---------|------------|---|---------|------------|
+   INT    |     X bits      |  X bits |     INT    |...|  X bits |     INT    |
+
+Note that addresses are only 21 bytes. That's the result of decoding the base58 addresses into binary representation and stripping the last 4 bytes which are just a checksum.
+
+Due to the high bandwidth of transactions created by MPMA sends, and to prevent spam from a given address to DDoS the counterparty network, each address can only transmit an MPMA send once per block. After and address transmits an MPMA send on a given block, subsequent MPMA sends from that same address will be ignored until the next block. This should also be reported as an error from the counterparty-lib client and RPC interfaces.
```
Comment (deweller) [inline cip-0010.md]: I disagree with this as it is difficult to predict when blocks confirm.  I think there are better ways to get the spam control we are looking for.

More discussion here:
https://counterpartytalk.org/t/cip0010-multi-peer-multi-asset-sends/2799/19?u=deweller

File: cip-0010.md (hunk 1 above)
Comment (chiguireitor) [reply cip-0010.md]: Yup, with network prefix without checksum like on enhanced sends

File: cip-0010.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+The MPMA sends messages should by identified by the message type 3 inside a standard, CIP11 compliant, message envelope.
+
+#### Example
+
+The following example is broken into several newlines, but the data stream is continuous on the encoded TX data. Numbers are preceded by a marker, 'x' for hexadecimal numbers, 'b' for bits.
+
+ * `x2000` LUT Size: 2 addresses, 1 bit to represent each one
```
[the remark this answers is not in the document]
Comment (chiguireitor) [reply cip-0010.md]: It's in little endian. Fixed on commit #a7a581d as robby's suggestion

File: cip-0010.md (hunk 3 above)
Comment (chiguireitor) [reply cip-0010.md]: See above

File: cip-0010.md (hunk 4 above)
Comment (chiguireitor) [reply cip-0010.md]: Good catch

File: cip-0010.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+ * `x00F3A6B6E4A093E5A5B9DA76977A5270FD4D62553E` Base58 decoded address (into hex, 21 bytes), LUT index 0
+ * `x009BDE9D244E7ABD061B0DCC1DFA8C0D880CD7DA21` Base58 decoded address (into hex, 21 bytes), LUT index 1
+ * `b1` Bit, Valid asset after this
+ * `x0100000000000000` 64 bits XCP asset ID
+ * `b1` 2 recipients
+ * `b0` Address 0 into the lookup table
+ * `x0000000001000000` Send 1 of the current Asset (XCP, divisible)
+ * `b1` Address 1 into the lookup table
+ * `x0000000004000000` Send 4 of the current Asset (XCP)
```
[the remark this answers is not in the document]
Comment (chiguireitor) [reply cip-0010.md]: Same here

File: cip-0010.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (deweller) [reply cip-0010.md]: Please rewrite these to standard big-endian format (`0x0002`).  Counterparty uses big-endian format for data encoding.

File: cip-0010.md (hunk 3 above)
Comment (deweller) [reply cip-0010.md]: Same as above

**Comment by chiguireitor** [comment]
Gonna close this PR as it includes unwanted test files. Fixes are coming in another PR.

File: cip-0010.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (chiguireitor) [reply cip-0010.md]: Okey dokey

**Pull request closed without merging**